docs/design/mono/wasm-threads.md

```diff
@@ -0,0 +1,158 @@
+ - allow HTTP and WS C# APIs to be used from any thread
+     - Underlying JS object have thread affinity
+ - don't change/break single threaded build. †
+ - don't try to block on UI thread.
+
+<sub><sup>† Note: all the text below discusses MT build only, unless explicit about ST build.</sup></sub>
+
+## Design proposal TL;DR - Alternative 10
+20. MAUI/BlazorWebView ... is the same thing
+21. execute whole WASM runtime on a worker. Blocking is fine there.
+22. The UI thread will only have `blazor.server.js` with small modifications.
+23. This needs new marketing name! -> `WASM server` ???
+
+# Detailed design
+
+## Blazor - what changes vs MAUI server
+- it still needs to compile for WASM target
+- it will have threadpool and blocking C# `.Wait` and `lock()`
+- it would not have Socket or DB connection
+
+## Blazor - what changes vs "Blazor WASM"
+- "onClick" would be asynchronous, same way as in Blazor server
+- Blazor's [`IJSInProcessRuntime.Invoke`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.jsinterop.ijsinprocessruntime.invoke) would not be available
+- Blazor's [`IJSUnmarshalledRuntime`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.jsinterop.ijsunmarshalledruntime) would not be available
+- no JavaScript APIs on `globalThis.Blazor` or `globalThis.MONO`
+- no runtime JS interop
+- no runtime on UI thread
+- Blazor united: progressive switch from remote server to WASM server, after WASM binaries arrived
+
+## Blazor startup
+- we will need to start emscripten on a worker and `postMessage` there all the configuration.
+    - this is possibly the biggest effort
+- probably no `Blazor.start`, `start.withResourceLoader` with returning `Response` object
+- we may keep `dotnet.js` loader to run in the UI
+    - TODO: prototype
+
+### Blazor `renderBatch`
+- streaming bytes - as Blazor server does
+    - we use [RenderBatchWriter](https://github.com/dotnet/aspnetcore/blob/045afcd68e6cab65502fa307e306d967a4d28df6/src/Components/Shared/src/RenderBatchWriter.cs) in the WASM
+    - we use `blazor.server.js` to render it
+    - preferred if fast enough
+    - could be ok to render batch by another thread while UI is rendering
+- plan b) use existing `SharedArrayBuffer`
+    - WASM would:
+        - stop the GC
+        - send the "read-memory-now"
+        - await for "done" message
+        - enable GC
+    - UI could duplicate `conv_string` and `unbox_mono_obj` dark magic and use it
+
+## Blazor JS interop
+- will work as usual on "Blazor server"
+
+## runtime JS Interop
+- would be available only to workers, but it would not have access to UI thread and DOM
+- could call sync and async calls both directions, promises and callbacks
+
+## HTTP & WS
+- would be running on a worker, will be accessible from all C# threads
+    - we will overcome thread affinity internally
+
+## C# Thread
+- could block on synchronization primitives
+- without JS interop. calling JSImport will PNSE.
```

Comment:
I think leaving out JSImport makes sense as a 1.0 version of this, but it's meaningful to support JSImport so that the "server" here can utilize browser APIs, to compensate for the lack of native APIs. Like for example if the "server" is able to use an offscreen canvas it would be able to access hardware accelerated rendering, which would enable developers to use stuff like browser machine learning APIs.

Reply:
To be able to do that we need JSImport/JSExport interop on the UI thread.

There are 3 choices how to do that
A) dispatch via managed as described in [deput-worker proposal](https://github.com/dotnet/runtime/pull/91696)
B) dispatch via just JS, which means we will have JS proxy in UI thread of C# proxy in side-car worker. [Comlink ](https://github.com/GoogleChromeLabs/comlink) style. We could write out own and make in also spin-blocking the UI. But this is double dispatch on each call. Hopping over 2 threads and their main loop. 
C) dispatch just the `mono_wasm_bind_js_function` and `mono_wasm_bind_cs_function` and then do the JS side of the marshaling in the UI thread. But the implementation is heavily dependent on 
- memory (that's easy one)
- shared code `invoke-cs.ts`, `invoke-js.ts` etc
- but that is dependent on Mono C methods and emscripten methods. Most of them are synchronous. 
   - emscripten: stack alllocation, memory views (growing)
   - GC and JS handles (I guess those should be UI thread local)
   - various JS helpers (logging, exception handling, asserts)
   - mono: string marshaling, gc roots
   - mono: call dispatch to managed code: instantiate `TaskCompletionSource` etc

Reply:
D) we could further move the boundary and have it between full `dotnet.runtime.js` on the UI thread and emscripten on the side-car. That would narrow it down to
- memory view events
- stackalloc/alloc/free
- mono_wasm_assembly_find_class
- mono_wasm_assembly_find_method
- mono_wasm_deregister_root
- mono_wasm_invoke_method_bound
- mono_wasm_invoke_method_ref
- mono_wasm_register_root
- mono_wasm_runtime_run_module_cctor

Reply:
I was proposing that the sidecar worker be able to access JS inside the sidecar context, to be clear. Not JS objects from the main thread. No remoting, so it can still be synchronous.

Reply:
We need some JS interop anyway for Blazor, the current surface is

- INTERNAL.loadLazyAssembly
- INTERNAL.loadSatelliteAssemblies
- Blazor._internal.dotNetCriticalError
- Blazor._internal.renderBatch
- globalThis.console.debug
- globalThis.console.error
- globalThis.console.info
- globalThis.console.warn
- Blazor._internal.endInvokeDotNetFromJS
- Blazor._internal.getApplicationEnvironment
- Blazor._internal.getPersistedState
- Blazor._internal.invokeJSJson
- Blazor._internal.receiveByteArray
- receiveHotReloadAsync
- MONO.getI16
- MONO.getI32
- MONO.getF32
- BINDING.js_string_to_mono_string
- BINDING.unbox_mono_obj
- BINDING.conv_string
- ICall InvokeJS